### PR TITLE
Transitions Emission Models

### PR DESCRIPTION
**Pull request by alexmaragko**
- Updated transitions.py emission models to include calculations for the "star", "isrf", "stellar_model", "approximate", and "convolved", keywords and their combinations.

- Updated mcfitted.py plotting method to match fitted.py.

- Updated tests.

Pending:
- Update transitions.py multiprocessing.
- Write "stellar_model" tests and provide fits file for the Kurucz stellar model.

### CODE REVIEW

File: amespahdbpythonsuite/mcfitted.py [hunk 1]
```diff
@@ -153,6 +154,11 @@ def plot(self, **keywords):
         """
         from astropy.nddata import StdDevUncertainty  # type: ignore
 
+        if keywords.get("datalabel", False):
```
Comment (PAHdb) [inline amespahdbpythonsuite/mcfitted.py]: `datalabel = keywords.get("datalabel", "obs")`

File: amespahdbpythonsuite/tests/test_transitions.py [hunk 2]
```diff
@@ -110,34 +110,81 @@ def test_shift(self, pahdb_theoretical):
...
+    )
     def test_partial_cascade(self, pahdb_theoretical):
         trans_multi = pahdb_theoretical.gettransitionsbyuid([18])
         data = trans_multi.get()
-        intf, tmax = transitions.Transitions._cascade_em_model(6 * 1.603e-12, data['data'][18])
+        intf, tmax = transitions.Transitions._cascade_em_model(
+            6 * 1.603e-12, data['data'][18]
+        )
         test_i = [x for x in intf if x["frequency"] == 3068.821][0]
         assert tmax == 1279.7835033561428
```
Comment (PAHdb) [inline amespahdbpythonsuite/tests/test_transitions.py:188]: Do `assert tmax` before assigning test_i.

File: amespahdbpythonsuite/transitions.py [hunk 3]
```diff
@@ -243,33 +312,72 @@ def calculatedtemperature(self, e: float, **keywords) -> None:
...
+                    Transitions.mean_energy_squared(**keywords) - energy**2
+                )
+
+            else:
+                energy = e
+                self.model['energy'][uid]['sigma'] = 0.0
+
+            self.model['energy'][uid]['e'] = energy
+
             print(
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:355]: Use `message` here.

File: amespahdbpythonsuite/transitions.py [hunk 4]
```diff
@@ -243,33 +312,72 @@ def calculatedtemperature(self, e: float, **keywords) -> None:
...
+                    Transitions.approximate_attained_temperature, 2.73, 5000.0
+                )
+            else:
+                Tmax = optimize.brentq(Transitions.attained_temperature, 2.73, 5000.0)
 
-            print("MAXIMUM ATTAINED TEMPERATURE     : %f Kelvin" % Tmax)
+            self.model["temperatures"][uid] = Tmax
 
-            self.model["temperatures"].append({"uid": uid, "temperature": Tmax})
+            print("MAXIMUM ATTAINED TEMPERATURE     : %f Kelvin" % Tmax)
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:373]: Use `message` here.

File: amespahdbpythonsuite/transitions.py [hunk 5]
```diff
@@ -337,11 +618,48 @@ def cascade(self, e: float, **keywords) -> None:
             i = 0
             nuids = len(self.uids)
             for uid in self.uids:
+                # Instatiate model energy and temperature dictionaries.
+                self.model['energy'][uid] = {}
+
                 print("SPECIES                          : %d/%d" % (i + 1, nuids))
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:624]: Use `message` here.

File: amespahdbpythonsuite/transitions.py [hunk 6]
```diff
@@ -337,11 +618,48 @@ def cascade(self, e: float, **keywords) -> None:
...
                 print("UID                              : %d" % uid)
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:625]: Use `message` here.

File: amespahdbpythonsuite/transitions.py [hunk 7]
```diff
@@ -337,11 +618,48 @@ def cascade(self, e: float, **keywords) -> None:
...
+                    if keywords.get('convolved'):
+                        Nphot = Transitions.number_of_photons(**keywords)
+
+                else:
+                    energy = e
+                    self.model['energy'][uid]['sigma'] = 0.0
+
+                self.model['energy'][uid]['e'] = energy
+
                 print(
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:659]: Yes, use `message` here.

File: amespahdbpythonsuite/transitions.py [hunk 8]
```diff
@@ -350,24 +668,54 @@ def cascade(self, e: float, **keywords) -> None:
                 global intensities
                 intensities = np.array([d["intensity"] for d in self.data[uid]])
 
-                Tmax = optimize.brentq(self.attainedtemperature, 2.73, 5000.0)
+                if keywords.get('approximate'):
+                    totalcrossection = np.sum(intensities)
 
+                Tmax = optimize.brentq(func1, 2.73, 5000.0)
+                self.model["temperatures"][uid] = Tmax
                 print("MAXIMUM ATTAINED TEMPERATURE     : %f Kelvin" % Tmax)
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:676]: `message`.

File: amespahdbpythonsuite/transitions.py [hunk 9]
```diff
@@ -350,24 +668,54 @@ def cascade(self, e: float, **keywords) -> None:
...
+                )
+            else:
+                intensities = np.array(
+                    [
+                        x * 2.48534271218563e-23 * nc * y**3 / totalcrossection
+                        for x, y in zip(intensities, frequencies)
+                    ]
+                )
+
+            print(
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:713]: `message`.

File: amespahdbpythonsuite/transitions.py [hunk 10]
```diff
@@ -350,24 +668,54 @@ def cascade(self, e: float, **keywords) -> None:
...
+                "ENERGY CONSERVATION IN SPECTRUM "
+                f" :{np.sum(intensities) / self.model['energy'][uid]['e']}"
+            )
+
+        print(57 * "=")
 
         elapsed = timedelta(seconds=(time.perf_counter() - tstart))
         print(f"Elapsed time: {elapsed}\n")
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:721]: Use `message` here. Of course, will need to have a better progress monitor at some point.

File: amespahdbpythonsuite/transitions.py [hunk 11]
```diff
@@ -543,7 +891,473 @@ def plot(self, **keywords) -> None:
...
+    def heat_capacity(T: float) -> float:
+        """
+        Calculate heat capacity.
+
+        :param T: Excitation temperature in Kelvin.
+        :type T: float
+
+        """
+
+        global frequencies
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py:1219]: Will need to find a mechanism to not use globals; either via a static class variable or using partials.

File: amespahdbpythonsuite/transitions.py [hunk 12]
```diff
@@ -563,41 +1377,104 @@ def featurestrength(T: float) -> float:
...
+                a = 1.8e-16
+                b = 8.1860
+
+        val = 1.4387751297850830401 * frequency / T
+
+        if (a == 0.0) or (val > np.log(np.finfo(float).max)):
+            return 0.0
+
+        else:
+            return 1 / ((np.exp(val) - 1) * a * T**b)
```
Comment (PAHdb) [inline amespahdbpythonsuite/transitions.py]: `np.exp` -> `np.expm1`.

**Review by PAHdb — CHANGES_REQUESTED** [review]
Looks great, few minor changes needed. I'm mostly presuming the emission model details are correct; the tests pass and cross comparison with IDL are favorable.

File: amespahdbpythonsuite/transitions.py (hunk 4 above)
Comment (alexmaragko) [reply amespahdbpythonsuite/transitions.py:373]: This is in the loop for each uid, so calling message will also generate two rows of '=' printed. Is this ok? This also applies to all cases within a loop for each uid in the following PR messages.

**Review by PAHdb — APPROVED** [review]
Looks good!

**Pull request merged**